cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_impl_dpc.cpp

```diff
@@ -621,485 +127,508 @@ train_splitter_impl<Float, Bin, Index, Task, use_private_mem>::compute_random_sp
+            const Index local_id = item.get_local_id(0);
 
+            const Index row_ofs = node_ptr[impl_const_t::ind_ofs];
+            const Index row_count = node_ptr[impl_const_t::ind_lrc];
             split_smp_t sp_hlp;
-
-            Index bs_ftr_bin = impl_const_t::leaf_mark_;
-            Index bs_ftr_id = impl_const_t::leaf_mark_;
-            Index bs_left_count = 0;
-
-            hist_type_t* local_hist_buf_ptr = local_hist_buf.get_pointer().get();
-
-            hist_type_t* bs_left_hist = nullptr;
-
-            hist_type_t prv_bs_left_hist[buff_size] = { hist_type_t(0) };
-            if constexpr (use_private_mem) {
-                bs_left_hist = &prv_bs_left_hist[0];
+            // Check node impurity
+            if (!sp_hlp.is_valid_impurity(node_imp_list_ptr, node_id, imp_threshold, row_count)) {
+                return;
             }
-            else {
-                bs_left_hist =
-                    fill_zero(local_hist_buf_ptr + (sub_group_local_id * 2 + 0) * class_count,
-                              class_count);
+            split_info_t global_bs;
+            if (local_id == 0) {
+                global_bs.init_clear(best_split_hist.get_pointer().get(), hist_prop_count);
             }
-
-            Float bs_left_imp = Float(0);
-            Float bs_imp_dec = min_imp_dec;
-
-            for (Index ftr_idx = sub_group_local_id; ftr_idx < selected_ftr_count;
-                 ftr_idx += sub_group_size) {
-                const Index ts_ftr_id =
-                    selected_ftr_list_ptr[node_id * selected_ftr_count + ftr_idx];
-                const Index ts_ftr_bin_count =
-                    bin_offset_list_ptr[ts_ftr_id + 1] - bin_offset_list_ptr[ts_ftr_id];
-
-                const hist_type_t* node_hist_ptr =
-                    node_hist_list_ptr +
-                    node_idx * selected_ftr_count * max_bin_count_among_ftrs * hist_prop_count;
-                const hist_type_t* ftr_hist_ptr =
-                    node_hist_ptr + ftr_idx * max_bin_count_among_ftrs * hist_prop_count;
-
-                Index ts_left_count = 0;
-                Index ts_right_count = 0;
-
-                hist_type_t* ts_left_hist = nullptr;
-                hist_type_t prv_ts_left_hist[buff_size] = { hist_type_t(0) };
-                if constexpr (use_private_mem) {
-                    ts_left_hist = &prv_ts_left_hist[0];
+            hist_type_t* const local_hist_ptr = local_hist.get_pointer().get();
+            hist_type_t* const buf_hist_ptr = buf_hist.get_pointer().get();
+            hist_type_t* const last_bin = last_bin_prev_batch.get_pointer().get();
+            split_scalar_t* const scalars_buf_ptr = scalars_buf.get_pointer().get();
+            Index* const last_bin_idx_ptr = last_bin_index.get_pointer().get();
+            // Due to local memory limitations need to proccess smaller blocks of features
+            for (Index batch_idx = 0; batch_idx < all_bin_count; batch_idx += batch_size) {
```

Comment:
I think this 200-line for loop that uses local memory to compute histogram needs additional explanations. I.e. how the global data is distributed; which parts of the global data are related to which local chunk, etc.